pennylane_lightning/src/algorithms/JacobianProd.hpp

```diff
@@ -122,48 +93,47 @@ class VectorJacobianProduct : public AdjointJacobian<T> {
-     * calculation.
-     * @param apply_operations Indicate whether to apply operations to psi prior
-     * to calculation.
+     * @param num_params Total number of parameters in the QuantumTape
+     * @param apply_operations Indicate whether to apply operations to jd.psi
+     * prior to calculation.
+     *
+     * @return std::function<std::vector<T>(const JacobianData<T> &jd)>
+     * where `jd` is a JacobianData object representing the QuantumTape
+     * to differentiate.
+     *
      */
-    void vectorJacobianProduct(std::vector<T> &vjp,
-                               std::vector<std::vector<T>> &jac,
-                               const std::vector<T> &dy,
-                               const std::complex<T> *psi, size_t num_elements,
-                               const std::vector<ObsDatum<T>> &observables,
-                               const OpsData<T> &operations,
-                               const std::vector<size_t> &trainableParams,
-                               bool apply_operations = false) {
-        const size_t num_params = trainableParams.size();
-
-        if (num_params == 0U || dy.empty()) {
-            vjp.clear();
-            return;
+    auto vectorJacobianProduct(const std::vector<T> &dy, size_t num_params,
+                               bool apply_operations = false)
+        -> std::function<std::vector<T>(const JacobianData<T> &)> {
+        if (dy.empty() ||
+            std::all_of(dy.cbegin(), dy.cend(), [](T e) { return e == 0; })) {
+            // If the dy vector is zero, then the
+            // corresponding element of the VJP will be zero,
+            // and we can avoid unnecessary computation.
+            return
+                [&num_params]([[maybe_unused]] const JacobianData<T> &jd)
+                    -> std::vector<T> { return std::vector<T>(num_params, 0); };
         }
 
-        const bool allzero =
-            std::all_of(dy.cbegin(), dy.cend(), [](T e) { return e == 0; });
-        if (allzero) {
-            vjp.resize(num_params);
-            return;
-        }
+        return [=](const JacobianData<T> &jd) -> std::vector<T> {
+            if (!jd.hasTrainableParams()) {
+                // The jd has no trainable parameters;
+                // the VJP is simple {}.
+                return {};
+            }
+
+            std::vector<T> vjp(num_params);
+            std::vector<T> jac(jd.getNumObservables() * num_params, 0);
 
-        this->adjointJacobian(psi, num_elements, jac, observables, operations,
-                              trainableParams, apply_operations);
+            // Compute Jacobian for the input jd using `adjoint` method
+            this->adjointJacobianJD(jac, jd, apply_operations);
 
-        computeVJP(vjp, jac, dy);
+            // Compute VJP
+            computeVJP(vjp, jac, dy, jd.getNumObservables(), num_params);
+            return vjp;
+        };
     }
 }; // class VectorJacobianProduct
```

Comment:
I also found this class derives `AdjointJacobian<T>`. Is it necessary? If so, we need to add a virtual destructor in AdjointJacobian<T>.

Reply:
Great catch! This is part of my first design decision for this class and the need for `createOpsData` method from `AdjointJacobian<T>`. Both classes are now updated.